lib/tcp.c

```diff
@@ -0,0 +1,220 @@
+#ifdef USE_NODELAY
+			int yes = 1;
+
+			/* Try disabling Nagle since we're latency-sensitive, UDP would
+			 * really be more appropriate but that's a much bigger change.
+			 */
+			(void) setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, (void *)&yes, sizeof(yes));
+#endif
+			break;
+		}
+
+		closesocket(sock);
+		sock = INVALID_SOCKET;
+	}
+
+#ifdef USE_GETADDRINFO
+	freeaddrinfo(addr);
+#endif
+
+	return sock;
+}
+
+static int sync_tcp_poll(void *ctxt, int *res_readable, int *res_writeable)
+{
+	struct sync_tcp *tcp = ctxt;
+#ifdef GEKKO
+	// libogc doesn't impmelent select()...
+	struct pollsd sds[1];
+	int events = (res_readable ? POLLIN : 0) | (res_writeable ? POLLOUT : 0)
+	sds[0].socket  = tcp->sock;
+	sds[0].events  = events;
+	sds[0].revents = 0;
+	if (net_poll(sds, 1, 0) < 0) return 0;
+	if (res_readable)
+		*res_readable = (sds[0].revents & POLLIN) && !(sds[0].revents & (POLLERR|POLLHUP|POLLNVAL));
+	if (res_writeable)
+		*res_writeable = (sds[0].revents & POLLOUT) && !(sds[0].revents & (POLLERR|POLLHUP|POLLNVAL));
+	return 1;
+#else
+	struct timeval to = { 0, 0 };
+	fd_set rfds, wfds;
+	int r;
+
+	FD_ZERO(&rfds);
+	FD_ZERO(&wfds);
+
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable: 4127)
+#endif
+	if (res_readable)
+		FD_SET(tcp->sock, &rfds);
+	if (res_writeable)
+		FD_SET(tcp->sock, &wfds);
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
+
+	r = select((int)tcp->sock + 1, &rfds, &wfds, NULL, &to);
+	if (r > 0) {
+		if (res_readable)
+			*res_readable = FD_ISSET(tcp->sock, &rfds);
+		if (res_writeable)
+			*res_writeable = FD_ISSET(tcp->sock, &wfds);
+	}
+	return r;
+
+#endif
+}
+
+static int sync_tcp_send(void *ctxt, const void *buf, int len)
+{
+	struct sync_tcp *tcp = ctxt;
+
+#ifdef WIN32
+	assert(len <= INT_MAX);
+	return send(tcp->sock, (const char *)buf, (int)len, 0);
+#else
+	return send(tcp->sock, (const char *)buf, len, 0);
+#endif
+}
+
+static int sync_tcp_recv(void *ctxt, void *buf, int len)
+{
+	struct sync_tcp *tcp = ctxt;
+
+#ifdef WIN32
+	assert(len <= INT_MAX);
+	return recv(tcp->sock, (char *)buf, (int)len, 0);
+#else
+	return recv(tcp->sock, (char *)buf, len, 0);
+#endif
+}
+
+static void sync_tcp_close(void *ctxt)
+{
+	struct sync_tcp *tcp = ctxt;
+
+	closesocket(tcp->sock);
+	free(tcp);
+}
+
+static struct sync_sockio_cb sync_tcp_sockio = {
+	.poll = sync_tcp_poll,
+	.send = sync_tcp_send,
+	.recv = sync_tcp_recv,
+	.close = sync_tcp_close,
+};
+
+int sync_tcp_connect(struct sync_device *d, const char *host, unsigned short port)
+{
+	struct sync_tcp *tcp;
+
+	tcp = malloc(sizeof(*tcp));
+	if (!tcp)
+		return -1;
+
+	tcp->sock = server_connect(host, port);
+	if (tcp->sock == INVALID_SOCKET) {
+		free(tcp);
+		return -1;
+	}
+
+	return sync_set_sockio_cb(d, &sync_tcp_sockio, tcp);
+}
+
+int sync_connect(struct sync_device *d, const char *host, unsigned short port)
+{
+	return sync_tcp_connect(d, host, port);
+}
+
+void sync_tcp_device_dtor(void)
+{
+	/* XXX: this is janky */
```

Comment:
How so? :)

Reply:
I'm pretty sure what I was thinking was "technically there's nothing preventing a single process from creating multiple sync_device instances simultaneously, but as-is this simplistic approach could wipe out socket_base on the first destroy (for USE_AMITCP), while another sync_device already exists, but now the socket library is unloaded..."

The code is just assuming That Won't Happen, without asserting to prevent it, or otherwise just handling it gracefully with a refcount or something.